pkg/websocket/connection.go

```diff
@@ -0,0 +1,167 @@
+limitations under the License.
+*/
+
+package websocket
+
+import (
+	"bytes"
+	"encoding/gob"
+	"errors"
+	"io"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/gorilla/websocket"
+)
+
+var (
+	// ErrConnectionNotEstablished is returned by methods that need a connection
+	// but no connection is already created.
+	ErrConnectionNotEstablished = errors.New("connection has not yet been established")
+)
+
+// RawConnection is an interface defining the methods needed
+// from a websocket connection
+type RawConnection interface {
+	WriteMessage(messageType int, data []byte) error
+	NextReader() (int, io.Reader, error)
+	Close() error
+}
+
+// ManagedConnection represents a websocket connection.
+type ManagedConnection struct {
+	connFactory    func() (RawConnection, error)
+	connection     RawConnection
+	connectionLock sync.RWMutex
+	messageBuffer  *bytes.Buffer
+	messageEncoder *gob.Encoder
+	closeChan      chan struct{}
+}
+
+// NewDurableSendingConnection creates a new websocket connection
+// that can only send messages to the endpoint it connects to.
+// The connection will continuously be kept alive and reconnected
+// in case of a loss of connectivity.
+func NewDurableSendingConnection(connFactory func() (RawConnection, error)) *ManagedConnection {
+	conn := newConnection(connFactory)
+
+	// Keep the connection alive asynchronously and reconnect on
+	// connection failure.
+	go func() {
+		// If the close signal races the connection attempt, make
+		// sure the connection actually closes.
+		defer func() {
+			conn.connectionLock.RLock()
+			defer conn.connectionLock.RUnlock()
+
+			if conn.connection != nil {
+				conn.connection.Close()
+			}
+		}()
+		for {
+			select {
+			default:
+				if err := conn.connect(); err != nil {
+					continue
+				}
+				conn.keepalive()
+			case <-conn.closeChan:
+				return
+			}
+		}
+	}()
+
+	return conn
+}
+
+// newConnection creates a new connection primitive.
+func newConnection(connFactory func() (RawConnection, error)) *ManagedConnection {
+	buffer := &bytes.Buffer{}
+	conn := &ManagedConnection{
+		connFactory:    connFactory,
+		messageBuffer:  buffer,
+		messageEncoder: gob.NewEncoder(buffer),
+		closeChan:      make(chan struct{}, 1),
+	}
+
+	return conn
+}
+
+// connect tries to establish a websocket connection.
+func (c *ManagedConnection) connect() (err error) {
+	wait.ExponentialBackoff(wait.Backoff{
+		Duration: 100 * time.Millisecond,
+		Factor:   1.3,
+		Steps:    20,
+		Jitter:   0.5,
+	}, func() (bool, error) {
+		conn, err := c.connFactory()
+		if err != nil {
+			return false, nil
+		}
+		c.connectionLock.Lock()
+		defer c.connectionLock.Unlock()
+
+		c.connection = conn
+		return true, nil
+	})
+
+	return err
+}
+
+// keepalive keeps the connection open and reads control messages.
+// All messages are discarded.
+func (c *ManagedConnection) keepalive() (err error) {
+	c.connectionLock.RLock()
+	defer c.connectionLock.RUnlock()
```

Comment:
I assume `c.connection` is a threadsafe implementation, which is why we only need `RLock` when calling through to it?